Src/Base/AMReX_TinyProfiler.cpp

```diff
@@ -488,9 +491,33 @@ TinyProfiler::MemoryFinalize (bool bFlushing) noexcept
     int ioproc = ParallelDescriptor::IOProcessorNumber();
     ParallelReduce::Max(dt_max, ioproc, ParallelDescriptor::Communicator());
 
-    for (std::size_t i=0; i<MemStat::memory_name.size(); ++i) {
+    for (std::size_t i=0; i<all_memstats.size(); ++i) {
         PrintMemStats(i, dt_max, t_final);
     }
+
+    if (!bFlushing) {
+        all_memstats.clear();
+        all_memnames.clear();
+    }
+}
+
+void
+TinyProfiler::RegisterArena (const std::string& memory_name,
+                             std::map<std::string, MemStat>& memstats) noexcept
+{
+    all_memstats.push_back(&memstats);
+    all_memnames.push_back(memory_name);
+}
+
+void
+TinyProfiler::DeregisterArena (std::map<std::string, MemStat>& memstats) noexcept
+{
+    for (std::size_t i = 0; i < all_memstats.size(); ++i) {
+        if (all_memstats[i] == &memstats) {
+            all_memstats.erase(all_memstats.begin() + i);
+            all_memnames.erase(all_memnames.begin() + i);
+        }
+    }
```

Comment:
In practice, this is OK because there is only one element in the vector that can meet the condition (unless one mistakenly calls RegisterArena twice).  If that's the case, we could used `std::find` and then `std::distance` to figure the element index first and then remove them from `all_memestats` and `all_memnames`.  We could also add an AMREX_ASSERT() in the end to make sure there are no more matches.  I think the code will be more readable.  The issue of the current code I cannot help noticing it will not work if there two matches in a row because of `++i` and `erase`.

Reply:
Good catch, thanks! This is intended to always work.